src/plugins/dashboard/public/application/listing/dashboard_unsaved_listing.tsx

```diff
@@ -0,0 +1,161 @@
+      <EuiFlexGroup alignItems="center" gutterSize="none">
+        <EuiFlexItem grow={false}>
+          <EuiIcon color="text" className="dshUnsavedListingItem__icon" type="dashboardApp" />
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiTitle size="xxs">
+            <h4 className="dshUnsavedListingItem__title">
+              {dashboard?.title ?? getNewDashboardTitle()}
+            </h4>
+          </EuiTitle>
+        </EuiFlexItem>
+      </EuiFlexGroup>
+      <EuiFlexGroup
+        gutterSize="none"
+        alignItems="flexStart"
+        className="dshUnsavedListingItem__actions"
+      >
+        <EuiFlexItem grow={false}>
+          <EuiButtonEmpty
+            flush="left"
+            size="s"
+            color="primary"
+            onClick={onOpenClick}
+            data-test-subj={`edit-unsaved-${title.split(' ').join('-')}`}
+            aria-label={dashboardUnsavedListingStrings.getEditAriaLabel(title)}
+          >
+            {dashboardUnsavedListingStrings.getEditTitle()}
+          </EuiButtonEmpty>
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiButtonEmpty
+            size="s"
+            color="danger"
+            onClick={onDiscardClick}
+            data-test-subj={`discard-unsaved-${title.split(' ').join('-')}`}
+            aria-label={dashboardUnsavedListingStrings.getDiscardAriaLabel(title)}
+          >
+            {dashboardUnsavedListingStrings.getDiscardTitle()}
+          </EuiButtonEmpty>
+        </EuiFlexItem>
+      </EuiFlexGroup>
+    </div>
+  );
+};
+
+export const DashboardUnsavedListing = ({ redirectTo }: { redirectTo: DashboardRedirect }) => {
+  const {
+    services: {
+      dashboardPanelStorage,
+      savedDashboards,
+      core: { overlays },
+    },
+  } = useKibana<DashboardAppServices>();
+
+  const [items, setItems] = useState<JSX.Element[]>([]);
+  const [dashboardIds, setDashboardIds] = useState<string[]>([]);
+
+  const onOpen = useCallback(
+    (id?: string) => {
+      redirectTo({ destination: 'dashboard', id, editMode: true });
+    },
+    [redirectTo]
+  );
+
+  const onDiscard = useCallback(
+    (id?: string) => {
+      confirmDiscardUnsavedChanges(overlays, () => {
+        dashboardPanelStorage.clearPanels(id);
+        setDashboardIds(dashboardPanelStorage.getDashboardIdsWithUnsavedChanges());
+      });
+    },
+    [overlays, dashboardPanelStorage]
+  );
+
+  useMount(() => {
+    setDashboardIds(dashboardPanelStorage.getDashboardIdsWithUnsavedChanges());
+  });
+
+  useEffect(() => {
+    let hasNewDashboard = false;
+    const dashPromises = dashboardIds
+      .filter((id) => {
+        if (id !== DASHBOARD_PANELS_UNSAVED_ID) {
+          return true;
+        }
+        hasNewDashboard = true;
+        return false;
+      })
+      .map((dashboardId) => savedDashboards.get(dashboardId));
+    Promise.all(dashPromises).then((dashboards: DashboardSavedObject[]) => {
```

Comment:
looks like these promises are causing a significant layout shift when the listing page is loaded. I wonder if it is worth having a spinner state with min-height that depends on a number of dashboardIds that we know before rendering.

Probably this could be especially annoying on slower network and because of this shift user may pick the wrong dashboard from a list

Reply:
Good call, I've added a loading state to the listing:

<img width="471" alt="Screen Shot 2021-01-29 at 2 45 38 PM" src="https://user-images.githubusercontent.com/14276393/106320667-2fd7cc00-6241-11eb-8ef3-93a1ffc4cf69.png">